[openstack] Nova API call optimizations

nmuesch opened the pull request:
<!-- *Note: Please remember to review the Datadog [Contribution Guidelines](https://github.com/DataDog/dd-agent/blob/master/CONTRIBUTING.md)
if you have not yet done so.* -->

### What does this PR do?

Uses the optional `changes-since` portion of the Nova API to not query a full list of things every time we hit the `/servers`/ endpoint. 

Additionally, we currently call the `servers` endpoint to get a list of all servers, then check `servers/{id}` to get some metadata about each individual server, then call the diagnostics endpoint. on that server if the server is in an active state.

Instead, we now call the `servers/detail` endpoint on the first check run, which will cache all metadata and server ids. We let nova filter this to only get servers in the `ACTIVE` state, to reduce payload sizes. Then we call the diagnostic method on all active servers. This removes the call to `server/{id}` entirely (which is called once for each server running)

On top of that, now we cache the list of servers, so that heavy call to `servers/details` only occurs on the first check run. Then we keep a timestamp of when we last called it, and on subsequent calls/check runs only get the list of servers that have had their state changed. 

### Motivation

We are making heavy calls to the Nova API, and a lot of them. Which causes the check to run long, and put a burden on the Openstack API when there are a large number of hosts. 

### Testing Guidelines

An overview on [testing](https://github.com/DataDog/dd-agent/blob/master/tests/README.md)
is available in our contribution guidelines.

### Versioning

- [ ] Bumped the check version in `manifest.json`
- [ ] Bumped the check version in `datadog_checks/{integration}/__init__.py`
- [ ] Updated `CHANGELOG.md`. Please use `Unreleased` as the date in the title
  for the new section.
- [ ] If PR impacts documentation, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new) 

### Additional Notes

Anything else we should know when reviewing?

ofek commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +850,60 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
-        url = '{0}/servers'.format(self.get_nova_endpoint())
+        # If we don't have a timestamp for this instance, default to None
+        if i_key not in self.changes_since_time:
+            self._changes_since_time[i_key] = None
```
Do you mean `self.changes_since_time[i_key]`?

ofek replied:
`self._changes_since_time[i_key]` vs
`self.changes_since_time[i_key]`

nmuesch replied:
You're right, updated.

ofek commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +850,60 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+            self._changes_since_time[i_key] = None
+
+        if self.changes_since_time[i_key]:
+            query_params['changes-since'] = changes_since_time
```
Where is `changes_since_time` declared?

ofek replied:
Then you'll needed to preface with `self.`

truthbk replied:
You could use a defaultdict, or `self.changes_since_time.get(i_key, None)`.

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +856,53 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+
+        # Since we ultimately only perform actions on ACTIVE servers
+        # Lets filter to only those to reduce the payload we receive
+        active_query_params['status'] = 'ACTIVE,DELETED,SHUTOFF'
```
```python
','.join(DIAGNOSTICABLE_STATES+REMOVED_STATES)
```

nmuesch replied:
It turns out that this isn't available for all versions of the API, so I'm making 2 addition API calls per check run to get the SHUTOFF and DELETED VMs. This shouldn't be an issue because its not called per server, so it shouldn't be a large overhead.

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -553,6 +558,13 @@ def __init__(self, name, init_config, agentConfig, instances=None):
         self.backoff = {}
         random.seed()
 
+        # Timestamp used to filter the call to get the list of nova servers
```
nit: It's not a timestamp but a an ISO8601 date time.

nmuesch replied:
Changed the comment :)

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +856,53 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+            new_server['project_name'] = self.get_project_name_from_id(new_server['tenant_id'])
+
+            for server in self.servers[i_key]:
+                if server_id == server['id'] and state not in DIAGNOSTICABLE_STATES:
```
You haven't initialized `server_id` or `state` anywhere... I think you mean `new_server['server_id']` and `new_server['state']`?

nmuesch replied:
Yupp, caught that while testing, thanks 🙇

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +856,53 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+            for server in self.servers[i_key]:
+                if server_id == server['id'] and state not in DIAGNOSTICABLE_STATES:
+                    self.servers[i_key].append(new_server)
+                elif server_id == server['id'] and state in REMOVED_STATES:
```
ditto

nmuesch replied:
done

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -553,6 +558,13 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+
+        # Cache the list of server id's available
+        # This will have a key of instance_id, and a list of servers
+        self.servers = {}
```
I think it's probably fine to 'split' the servers by instance id (although I'm not sure it's necessary - read ahead), but I think it would probably also make sure to index the servers by sever `id` or by `name`. This would allow you to get to a server's details in O(1), to verify state, etc.

ie. `self.servers[i_key][id]` or `self.servers[i_key][name]`. 

I understand that you need to keep track of the servers actually in use by each instance (ie. tenant/project), I'm wondering if it wouldn't be better for you to keep a separate cache, indexed by instance key, to a set of server ids available in a certain instance, and a separate cache for the actual server details. 

We should also ask ourselves if it makes sense to actually cache the servers by instance key (if a server can indeed belong to two different tenants or whatever, not sure it's possible). But even if it is possible, there's just the one instance which cannot be powered on and off at the same time. If we cache by instance then we'll have inconsistent states for the same VM in the cache. It's probably not the end of the world as we would get the right state for the VM in the next iteration. 

I just wanted to bring this up because it feels like it might be more efficient ways to index and iterate these caches which might get pretty bulky.

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -876,39 +917,27 @@ def get_project_name_from_id(self, tenant_id):
-        hypervisor_hostname = None
-        tenant_id = None
-        project_name = None
+        server_id = server.get('id')
```
do you mean `server_details.get('id')`?

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +856,53 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+                if server_id == server['id'] and state not in DIAGNOSTICABLE_STATES:
+                    self.servers[i_key].append(new_server)
+                elif server_id == server['id'] and state in REMOVED_STATES:
+                    self.servers[i_key][:] = [server for server in self.servers[i_key] if self.servers[i_key]('server_id') != server_id]
```
As mentioned earlier, I'm not a fan of huge fan of using lists here, it doesn't seem like the right data structure as these lists can get quite large.

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -876,39 +917,27 @@ def get_project_name_from_id(self, tenant_id):
+        project_name = self.get_project_name_from_id(tenant_id)
+
+        server_stats = {}
+        url = '{0}/servers/{1}/diagnostics'.format(self.get_nova_endpoint(), server_id)
```
If you decide to keep servers in the cache in `SHUTOFF` state, remember that we should only be calling this if `state` is in a `DIAGNOSTICABLE` state (not much to worry about as the code is resilient to `InstancePowerOffFailure`).

truthbk commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -95,6 +96,11 @@
 
+REMOVED_STATES = [
+    'DELETED',
+    'SHUTOFF'
```
unsure `SHUTOFF` should lead to the the same actions as `REMOVED`; I see you use these states to clear out `servers` from the cache.... I would ask myself if it's worth it to keep `SHUTOFF` servers around in the cache. I don't have a strong opinion, but would like us to think about it. This would allow us to report a metric on number of stopped VMs, etc... or ultimately add some "sensors" for stopped VM's whatever they may be. Bear in mind that VMs might have storage, or networking or whatever, associated with them (and although we might not be reporting those metrics now, it' might be worth it or more future proof to keep as rich a server registry in cache as we can). Then again, a larger registry could also add a performance penalty.

nmuesch replied:
So I think that having this would let us add some future metrics about server types and what not, but I don't think having them in the cache is worthwhile for this performance based work. 

In terms of being future proof, I think having the cache setup as it currently is, is rather extensible for the future should we want to add in those metrics.

nmuesch commented:
Ok, I believe I addressed most of the comments here. The last thing to look into is the value of getting servers that are SHUTOFF, which I'm currently looking into. Thanks for the feedback on this 🙇

ofek commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +858,64 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+            self.changes_since_time[i_key] = None
+
+        if self.changes_since_time[i_key]:
+            query_params['changes-since'] = self.changes_since_time[i_key]
```
You can remove these 4 lines and just do:

`query_params['changes-since'] = self.changes_since_time.get(i_key)`

ofek commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +858,62 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+            # Get a list of deleted serversTimestamp used to filter the call to get the list
+            # Need to have admin perms for this to take affect
+            query_params['deleted'] = 'true'
+            query_params['status'] = ''
```
Should this be an empty string or should it be `None` to omit it?

nmuesch commented:
I've added in a test for the caching system and also fixed the existing test since I changed some function names/return values.

ofek commented on openstack/datadog_checks/openstack/openstack.py:
```diff
@@ -844,24 +858,62 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
+            new_server['server_id'] = server.get('id')
+            new_server['state'] = server.get('status')
+            new_server['server_name'] = server.get('name')
+            new_server['hypervisor_hostname'] = server.get('OS-EXT-SRV-ATTR:hypervisor_hostname')
```
Should this be

```
new_server['server_id'] = server.get('server_id')
new_server['state'] = server_details.get('state')
new_server['server_name'] = server_details.get('server_name')
new_server['hypervisor_hostname'] = server_details.get('hypervisor_hostname')
```

as they are elsewhere?
